frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -5794,7 +5933,57 @@ CallResolutionResult resolveTheseCall(ResolutionContext* rc,
                      /* hasQuestionArg */ false,
                      /* isParenless */ false,
                      /* actuals */ std::move(actuals));
-  return resolveGeneratedCall(rc->context(), astContext, ci, inScopes);
+  auto cr = resolveGeneratedCall(rc->context(), astContext, ci, inScopes);
+  return callResolutionResultToTheseResolutionResult(cr, receiverType);
+}
+
+const types::QualifiedType& getPromotionType(Context* context, types::QualifiedType qt) {
+  QUERY_BEGIN(getPromotionType, context, qt);
+
+
+  auto ret = QualifiedType();
+  if (qt.isUnknownOrErroneous()) {
+    /* do nothing */
+  } else if (auto loopIt = qt.type()->toLoopExprIteratorType()) {
+    ret = loopIt->yieldType();
+  } else if (auto fnIt = qt.type()->toFnIteratorType()) {
+    ResolutionContext rc(context); // TODO
+    ret = yieldType(&rc, fnIt->iteratorFn(), fnIt->poiScope());
+  } else if (auto promoIt = qt.type()->toPromotionIteratorType()) {
+    ResolutionContext rc(context); // TODO
```

Comment:
A trivial resolution context is not correct in this situation, because the iterator could've been defined inside another function. The resolution context would normally contain this information, but at this point, there's no way of knowing. I'll expand on the TODOs and add to the future work.